File: app/src/main/java/fr/paug/androidmakers/ui/activity/SessionDetailActivity.java
package fr.paug.androidmakers.ui.activity;

import android.animation.ObjectAnimator;
import android.content.Context;
import android.content.Intent;
import android.databinding.DataBindingUtil;
import android.graphics.Color;
import android.graphics.drawable.AnimatedVectorDrawable;
import android.net.Uri;
import android.os.Build;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.annotation.RequiresApi;
import android.support.v4.content.ContextCompat;
import android.text.Html;
import android.text.TextUtils;
import android.text.format.DateUtils;
import android.text.method.LinkMovementMethod;
import android.util.Log;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.view.animation.AnimationUtils;

import java.util.ArrayList;
import java.util.Formatter;
import java.util.List;

import fr.paug.androidmakers.BuildConfig;
import fr.paug.androidmakers.R;
import fr.paug.androidmakers.databinding.ActivityDetailBinding;
import fr.paug.androidmakers.databinding.DetailViewSpeakerInfoElementBinding;
import fr.paug.androidmakers.databinding.SmallRibbonImageBinding;
import fr.paug.androidmakers.databinding.SmallSocialImageBinding;
import fr.paug.androidmakers.manager.AgendaRepository;
import fr.paug.androidmakers.model.Ribbon;
import fr.paug.androidmakers.model.Room;
import fr.paug.androidmakers.model.Session;
import fr.paug.androidmakers.model.SocialNetworkHandle;
import fr.paug.androidmakers.model.Speaker;
import fr.paug.androidmakers.ui.adapter.ScheduleSession;
import fr.paug.androidmakers.ui.util.CheckableFloatingActionButton;
import fr.paug.androidmakers.util.ScheduleSessionHelper;
import fr.paug.androidmakers.util.SessionSelector;
import fr.paug.androidmakers.util.UIUtils;

/**
 * Details of a session
 *
 * Nice improvements to have : video link, session rate/feedback
 */
public class SessionDetailActivity extends BaseActivity {

    private static final String PARAM_SESSION_ID = "param_session_id";
    private static final String PARAM_SESSION_START_DATE = "param_session_start_date";
    private static final String PARAM_SESSION_END_DATE = "param_session_end_date";
    private static final String PARAM_SESSION_ROOM = "param_session_room";

    private int sessionId;
    private long sessionStartDateInMillis;
    private long sessionEndDateInMillis;

    private Session session;
    private String sessionDateAndRoom;
    private List<String> speakersList = new ArrayList<>();

    public static void startActivity(Context context, ScheduleSession scheduleSession) {
        Intent intent = new Intent(context, SessionDetailActivity.class);
        intent.putExtra(PARAM_SESSION_ID, scheduleSession.getSessionId());
        intent.putExtra(PARAM_SESSION_START_DATE, scheduleSession.getStartTimestamp());
        intent.putExtra(PARAM_SESSION_END_DATE, scheduleSession.getEndTimestamp());
        intent.putExtra(PARAM_SESSION_ROOM, scheduleSession.getRoomId());
        context.startActivity(intent);
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        final ActivityDetailBinding activityDetailBinding = DataBindingUtil.setContentView(this, R.layout.activity_detail);
        sessionId = getIntent().getIntExtra(PARAM_SESSION_ID, -1);
        session = AgendaRepository.getInstance().getSession(sessionId);

        sessionStartDateInMillis = getIntent().getLongExtra(PARAM_SESSION_START_DATE, -1);
        sessionEndDateInMillis = getIntent().getLongExtra(PARAM_SESSION_END_DATE, -1);

        final Room sessionRoom = AgendaRepository.getInstance().getRoom(
                getIntent().getIntExtra(PARAM_SESSION_ROOM, -1)
        );

        if (session == null) {
            // We have a problem !
            activityDetailBinding.sessionInformationsScrollView.setVisibility(View.GONE);
            activityDetailBinding.errorMessageTextView.setVisibility(View.VISIBLE);
            return;
        }

        final String sessionDate = DateUtils.formatDateRange(this,
                new Formatter(getResources().getConfiguration().locale),
                sessionStartDateInMillis,
                sessionEndDateInMillis,
                DateUtils.FORMAT_SHOW_WEEKDAY | DateUtils.FORMAT_SHOW_TIME,
                null).toString();
        sessionDateAndRoom = sessionRoom != null &&
                !TextUtils.isEmpty(sessionRoom.name) ?
                getString(R.string.sessionDateWithRoomPlaceholder, sessionDate, sessionRoom.name) : sessionDate;

        activityDetailBinding.sessionTitleTextView.setText(session.title);
        activityDetailBinding.sessionDateAndRoomTextView.setText(sessionDateAndRoom);
        activityDetailBinding.sessionDescriptionTextView.setMovementMethod(LinkMovementMethod.getInstance());
        activityDetailBinding.sessionDescriptionTextView.setText(session.description != null ?
                Html.fromHtml(session.description.trim()) : "");

        final int languageFullNameRes = session.getLanguageName();
        if (languageFullNameRes != 0) {
            activityDetailBinding.sessionLanguageChip.setChipText(getString(languageFullNameRes));
        } else {
            activityDetailBinding.sessionLanguageChip.setVisibility(View.GONE);
        }

        if (session.subtype != null) {
            String capitalizedSubType = session.subtype.substring(0, 1).toUpperCase() + session.subtype.substring(1);
            activityDetailBinding.sessionSubTypeChip.setChipText(capitalizedSubType);
        } else {
            activityDetailBinding.sessionSubTypeChip.setVisibility(View.GONE);
        }

        if (session.type != null) {
            activityDetailBinding.sessionTypeChip.setChipText(session.type);
        } else {
            activityDetailBinding.sessionTypeChip.setVisibility(View.GONE);
        }

        if (session.experience != null) {
            activityDetailBinding.sessionExperienceChip.setChipText(session.experience);
        } else {
            activityDetailBinding.sessionExperienceChip.setVisibility(View.GONE);
        }

        final ViewGroup sessionSpeakerLayout = findViewById(R.id.sessionSpeakerLayout);
        if (session.speakers != null && session.speakers.length > 0) {
            for (final int speakerID : session.speakers) {
                final Speaker speaker = AgendaRepository.getInstance().getSpeaker(speakerID);
                if (speaker != null) {
                    speakersList.add(speaker.getFullNameAndCompany());

                    final DetailViewSpeakerInfoElementBinding speakerInfoElementBinding =
                            DataBindingUtil.inflate(getLayoutInflater(),
                                    R.layout.detail_view_speaker_info_element,
                                    null,
                                    false);
                    speakerInfoElementBinding.speakerBio.setMovementMethod(LinkMovementMethod.getInstance());
                    speakerInfoElementBinding.setSpeaker(speaker);

                    setSpeakerSocialNetworkHandle(speaker, speakerInfoElementBinding);
                    setSpeakerRibbons(speaker, speakerInfoElementBinding);

                    sessionSpeakerLayout.addView(speakerInfoElementBinding.getRoot());
                }
            }
        }

        activityDetailBinding.scheduleFab.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View fab) {
                boolean isInSchedule = !((CheckableFloatingActionButton) fab).isChecked();
                ((CheckableFloatingActionButton) fab).setChecked(isInSchedule);
                changeSessionSelection(isInSchedule);

                if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                    animateFab((CheckableFloatingActionButton) fab, isInSchedule);
                }
            }
        });

        final boolean sessionSelected = SessionSelector.getInstance().isSelected(sessionId);
        activityDetailBinding.scheduleFab.setChecked(sessionSelected);

        setActionBar(session);
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    void animateFab(CheckableFloatingActionButton fab, Boolean isInSchedule) {
        AnimatedVectorDrawable avd = (AnimatedVectorDrawable) ContextCompat.getDrawable(
                this, isInSchedule ? R.drawable.avd_bookmark : R.drawable.avd_unbookmark);
        fab.setImageDrawable(avd);
        ObjectAnimator backgroundColor = ObjectAnimator.ofArgb(
                fab,
                UIUtils.BACKGROUND_TINT,
                isInSchedule ? Color.WHITE
                        : ContextCompat.getColor(this, R.color.colorAccent));
        backgroundColor.setDuration(400L);
        backgroundColor.setInterpolator(AnimationUtils.loadInterpolator(this,
                android.R.interpolator.fast_out_slow_in));
        backgroundColor.start();
        avd.start();
    }

    private void setSpeakerSocialNetworkHandle(Speaker speaker, DetailViewSpeakerInfoElementBinding speakerInfoElementBinding) {
        if (speaker.socialNetworkHandles != null && speaker.socialNetworkHandles.size() > 0) {
            for (final SocialNetworkHandle socialNetworkHandle : speaker.socialNetworkHandles) {
                if (socialNetworkHandle.networkType != SocialNetworkHandle.SocialNetworkType.Unknown) {
                    final SmallSocialImageBinding smallSocialImageBinding = DataBindingUtil.inflate(getLayoutInflater(), R.layout.small_social_image, null, false);
                    smallSocialImageBinding.setSocialHandle(socialNetworkHandle);
                    smallSocialImageBinding.image.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View view) {
                            if (BuildConfig.DEBUG) {
                                Log.d(SessionDetailActivity.class.getName(), "User clicked on social handle with name=" + socialNetworkHandle.networkType.name());
                            }
                            try {
                                startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse(socialNetworkHandle.link)));
                            } catch (Exception e) {
                                e.printStackTrace();
                            }
                        }
                    });
                    speakerInfoElementBinding.speakerSocialNetworkHandleLayout.addView(smallSocialImageBinding.getRoot());
                }
            }
        } else {
            speakerInfoElementBinding.speakerSocialNetworkHandleLayout.setVisibility(View.GONE);
        }
    }

    private void setSpeakerRibbons(Speaker speaker, DetailViewSpeakerInfoElementBinding speakerInfoElementBinding) {
        if (speaker.ribbonList != null && speaker.ribbonList.size() > 0) {
            for (final Ribbon ribbon : speaker.ribbonList) {
                if (ribbon.ribbonType != Ribbon.RibbonType.NONE) {
                    final SmallRibbonImageBinding smallRibbonImageBinding =
                            DataBindingUtil.inflate(getLayoutInflater(), R.layout.small_ribbon_image, null, false);
                    smallRibbonImageBinding.setRibbon(ribbon);
                    smallRibbonImageBinding.image.setOnClickListener(new View.OnClickListener() {
                        @Override
                        public void onClick(View view) {
                            if (BuildConfig.DEBUG) {
                                Log.d(SessionDetailActivity.class.getName(), "User clicked on ribbon with name=" + ribbon.ribbonType.name());
                            }
                            try {
                                startActivity(new Intent(Intent.ACTION_VIEW, Uri.parse(ribbon.link)));
                            } catch (Exception e) {
                                e.printStackTrace();
                            }
                        }
                    });
                    speakerInfoElementBinding.speakerRibbonLayout.addView(smallRibbonImageBinding.getRoot());
                } else {
                    speakerInfoElementBinding.speakerRibbonLayout.setVisibility(View.GONE);
                }
            }
        } else {
            speakerInfoElementBinding.speakerRibbonLayout.setVisibility(View.GONE);
        }
    }

    private void setActionBar(Session session) {
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
            getSupportActionBar().setTitle(session.title);
        }
    }

    private void changeSessionSelection(boolean select) {
        SessionSelector.getInstance().setSessionSelected(sessionId, select);
        toggleScheduleSessionNotification(select);
    }

    private void toggleScheduleSessionNotification(boolean select) {
        if (select) {
            ScheduleSessionHelper.scheduleStarredSession(this, sessionStartDateInMillis, sessionEndDateInMillis, sessionId);
        } else {
            ScheduleSessionHelper.unScheduleSession(this, sessionId);
        }
    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.detail, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        switch (item.getItemId()) {
            case android.R.id.home:
                onBackPressed();
                return true;
            case R.id.share:
                shareSession();
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    private void shareSession() {
        String speakers = TextUtils.join(", ", speakersList);

        Intent shareSessionIntent = new Intent(Intent.ACTION_SEND);
        shareSessionIntent.putExtra(Intent.EXTRA_SUBJECT, session.title);
        shareSessionIntent.putExtra(Intent.EXTRA_TEXT,
                String.format("%s (%s, %s, %s)", session.title, speakers, sessionDateAndRoom, getString(session.getLanguageName())));
        shareSessionIntent.setType("text/plain");
        startActivity(shareSessionIntent);
    }

}

File: app/src/main/java/fr/paug/androidmakers/ui/activity/SessionDetailActivity.kt
package fr.paug.androidmakers.ui.activity

import android.animation.ObjectAnimator
import android.content.Context
import android.content.Intent
import android.graphics.Color
import android.graphics.drawable.AnimatedVectorDrawable
import android.net.Uri
import android.os.Build
import android.os.Bundle
import android.text.Html
import android.text.TextUtils
import android.text.format.DateUtils
import android.text.method.LinkMovementMethod
import android.util.Log
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import android.view.animation.AnimationUtils
import androidx.annotation.RequiresApi
import androidx.core.content.ContextCompat
import fr.paug.androidmakers.BuildConfig
import fr.paug.androidmakers.R
import fr.paug.androidmakers.databinding.ActivityDetailBinding
import fr.paug.androidmakers.databinding.DetailViewSpeakerInfoElementBinding
import fr.paug.androidmakers.databinding.SmallSocialImageBinding
import fr.androidmakers.store.manager.FirebaseStore
import fr.androidmakers.store.model.*
import fr.paug.androidmakers.ui.adapter.ScheduleSession
import fr.paug.androidmakers.ui.util.CheckableFloatingActionButton
import fr.paug.androidmakers.util.ScheduleSessionHelper
import fr.paug.androidmakers.util.SessionSelector
import fr.paug.androidmakers.util.UIUtils
import kotlinx.coroutines.*
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.combine
import kotlinx.coroutines.flow.first
import java.util.*

class SessionDetailActivity : BaseActivity() {

    private lateinit var activityDetailBinding: ActivityDetailBinding
    private var sessionId: String = ""
    private var sessionStartDateInMillis: Long = 0
    private var sessionEndDateInMillis: Long = 0

    private var sessionDateAndRoom: String? = null
    private val speakersList = ArrayList<String>()
    private var sessionForShare: Session? = null

    val scope = object : CoroutineScope {
        override val coroutineContext = Dispatchers.Main + Job()
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        activityDetailBinding = ActivityDetailBinding.inflate(layoutInflater)
        setContentView(activityDetailBinding.root)
        sessionId = intent.getStringExtra(PARAM_SESSION_ID)!!
        val roomId = intent.getStringExtra(PARAM_SESSION_ROOM)

        scope.launch {
            combine(
                    FirebaseStore().getSession(sessionId),
                    FirebaseStore().getRoom(roomId!!)
            ) { session, room ->
                session to room
            }.collect {
                setupUI(
                        it.first,
                        it.second,
                        roomId)

            }
        }
    }

    fun setupUI(session: Session, room: Room, roomId: String) {
        // small hack for share
        // Ideally, we should only create the menu when we have the data
        sessionForShare = session
        sessionStartDateInMillis = intent.getLongExtra(PARAM_SESSION_START_DATE, -1)
        sessionEndDateInMillis = intent.getLongExtra(PARAM_SESSION_END_DATE, -1)

        setupFeedback(session, roomId, sessionStartDateInMillis)

        val sessionDate = DateUtils.formatDateRange(this,
                Formatter(resources.configuration.locale),
                sessionStartDateInMillis,
                sessionEndDateInMillis,
                DateUtils.FORMAT_SHOW_WEEKDAY or DateUtils.FORMAT_SHOW_TIME or DateUtils.FORMAT_SHOW_DATE or DateUtils.FORMAT_SHOW_YEAR,
                null).toString()
        sessionDateAndRoom = if (!TextUtils.isEmpty(room.name))
            getString(R.string.sessionDateWithRoomPlaceholder, sessionDate, room.name)
        else
            sessionDate
        activityDetailBinding.sessionDateAndRoomTextView.text = sessionDateAndRoom

        activityDetailBinding.sessionTitleTextView.text = session.title
        activityDetailBinding.sessionDescriptionTextView.text = if (session.description != null)
            Html.fromHtml(session.description.trim { it <= ' ' })
        else
            ""

        val languageFullNameRes = session.language
        if (languageFullNameRes.isNotEmpty() == true) {
            activityDetailBinding.sessionLanguageChip.text = languageFullNameRes
            activityDetailBinding.sessionLanguageChip.visibility = View.VISIBLE
        }

        // Tags list
        if (session.tags.isNotEmpty() == true) {
            if (session.tags.first().isNotEmpty()) {
                activityDetailBinding.sessionTypeChip.text = session.tags.first()
                activityDetailBinding.sessionTypeChip.visibility = View.VISIBLE
            }
        }

        if (session.complexity.isNotEmpty() == true) {
            activityDetailBinding.sessionExperienceChip.text = session.complexity
            activityDetailBinding.sessionExperienceChip.visibility = View.VISIBLE
        }

        activityDetailBinding.scheduleFab.setOnClickListener { fab ->
            val isInSchedule = !(fab as CheckableFloatingActionButton).isChecked
            fab.isChecked = isInSchedule
            changeSessionSelection(isInSchedule)

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.LOLLIPOP) {
                animateFab(fab, isInSchedule)
            }
        }

        val sessionSelected = SessionSelector.isSelected(sessionId)
        activityDetailBinding.scheduleFab.isChecked = sessionSelected

        setActionBar(session)
        setSpeakers(session)

        if (session.platformUrl != null) {
            activityDetailBinding.watchButton.visibility = View.VISIBLE
            activityDetailBinding.watchButton.setOnClickListener {
                startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(session.platformUrl)))
            }
        }

        if (session.slido != null) {
            activityDetailBinding.slidoButton.visibility = View.VISIBLE
            activityDetailBinding.slidoButton.setOnClickListener {
                startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(session.slido)))
            }
        }

        activityDetailBinding.separator2.visibility = View.VISIBLE
    }

    private fun setupFeedback(session: Session, roomId: String, sessionStartDateInMillis: Long) {
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.detail, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean = when (item.itemId) {
        android.R.id.home -> {
            onBackPressed()
            true
        }
        R.id.share -> {
            sessionForShare?.let { shareSession(it) }
            true
        }
        else -> super.onOptionsItemSelected(item)
    }

    override fun onDestroy() {
        super.onDestroy()
        scope.cancel()
    }

    private fun setSpeakers(session: Session) {

        scope.launch(Dispatchers.Main) {
            val speakers = session.speakers.filter { !it.isNullOrBlank() }
                    .map {
                        async {
                            FirebaseStore().getSpeaker(it).first()
                        }
                    }
                    .awaitAll()
                    .filterNotNull()
            val sessionSpeakerLayout = findViewById<ViewGroup>(R.id.sessionSpeakerLayout)
            sessionSpeakerLayout.removeAllViews()

            speakers.forEach {speaker ->
                speakersList.add(speaker.getFullNameAndCompany())

                val speakerInfoElementBinding = DetailViewSpeakerInfoElementBinding.inflate(layoutInflater,
                        null,
                        false)
                speakerInfoElementBinding.speakerBio.movementMethod = LinkMovementMethod.getInstance()

                setSpeakerSocialNetworkHandle(speaker, speakerInfoElementBinding)
                sessionSpeakerLayout.addView(speakerInfoElementBinding.root)
            }
        }
    }

    @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
    internal fun animateFab(fab: CheckableFloatingActionButton, isInSchedule: Boolean) {
        (ContextCompat.getDrawable(
                this, if (isInSchedule) R.drawable.avd_bookmark else R.drawable.avd_unbookmark) as? AnimatedVectorDrawable)?.also { avd ->
            fab.setImageDrawable(avd)
            val backgroundColor = ObjectAnimator.ofArgb(
                    fab,
                    UIUtils.BACKGROUND_TINT,
                    if (isInSchedule)
                        Color.WHITE
                    else
                        ContextCompat.getColor(this, R.color.colorAccent))
            backgroundColor.duration = 400L
            backgroundColor.interpolator = AnimationUtils.loadInterpolator(this,
                    android.R.interpolator.fast_out_slow_in)
            backgroundColor.start()
            avd.start()
        }
    }

    private fun setSpeakerSocialNetworkHandle(speaker: Speaker, speakerInfoElementBinding: DetailViewSpeakerInfoElementBinding) {
        if (speaker.socials != null && speaker.socials!!.isNotEmpty()) {
            for (social in speaker.socials!!) {
                val socialNetworkHandle = SocialNetworkHandle(social?.icon, social?.link)
                if (socialNetworkHandle.networkType != SocialNetworkHandle.SocialNetworkType.Unknown) {
                    val smallSocialImageBinding = SmallSocialImageBinding.inflate(layoutInflater, null, false)
                    smallSocialImageBinding.image.setOnClickListener {
                        if (BuildConfig.DEBUG) {
                            Log.d(SessionDetailActivity::class.java.name, "User clicked on social handle with name=" + socialNetworkHandle.networkType)
                        }
                        try {
                            startActivity(Intent(Intent.ACTION_VIEW, Uri.parse(social?.link)))
                        } catch (e: Exception) {
                            e.printStackTrace()
                        }
                    }
                    speakerInfoElementBinding.speakerSocialNetworkHandleLayout.addView(smallSocialImageBinding.root)
                }
            }
        } else {
            speakerInfoElementBinding.speakerSocialNetworkHandleLayout.visibility = View.GONE
        }
    }

    private fun setActionBar(session: Session?) {
        if (supportActionBar != null) {
            supportActionBar?.setDisplayHomeAsUpEnabled(true)
            supportActionBar?.title = session?.title
        }
    }

    private fun changeSessionSelection(select: Boolean) {
        SessionSelector.setSessionSelected(sessionId, select)
        toggleScheduleSessionNotification(select)
    }

    private fun toggleScheduleSessionNotification(select: Boolean) {
        if (select) {
            ScheduleSessionHelper.scheduleStarredSession(this, sessionStartDateInMillis, sessionEndDateInMillis, sessionId)
        } else {
            ScheduleSessionHelper.unScheduleSession(this, sessionId)
        }
    }

    private fun shareSession(session: Session) {
        val speakers = TextUtils.join(", ", speakersList)

        val shareSessionIntent = Intent(Intent.ACTION_SEND)
        shareSessionIntent.putExtra(Intent.EXTRA_SUBJECT, session.title)
        if (speakersList.isEmpty()) {
            shareSessionIntent.putExtra(Intent.EXTRA_TEXT,
                    String.format("%s: %s (%s)", getString(R.string.app_name), session.title, sessionDateAndRoom))
        } else {
            shareSessionIntent.putExtra(Intent.EXTRA_TEXT,
                    String.format("%s: %s (%s, %s, %s)", getString(R.string.app_name), session.title, speakers, sessionDateAndRoom, session.language))
        }
        shareSessionIntent.type = "text/plain"
        startActivity(shareSessionIntent)
    }

    companion object {
        private const val PARAM_SESSION_ID = "param_session_id"
        private const val PARAM_SESSION_START_DATE = "param_session_start_date"
        private const val PARAM_SESSION_END_DATE = "param_session_end_date"
        private const val PARAM_SESSION_ROOM = "param_session_room"

        fun startActivity(context: Context, scheduleSession: ScheduleSession) {
            val intent = Intent(context, SessionDetailActivity::class.java)
            intent.putExtra(PARAM_SESSION_ID, scheduleSession.sessionId)
            intent.putExtra(PARAM_SESSION_START_DATE, scheduleSession.startTimestamp)
            intent.putExtra(PARAM_SESSION_END_DATE, scheduleSession.endTimestamp)
            intent.putExtra(PARAM_SESSION_ROOM, scheduleSession.roomId)
            context.startActivity(intent)
        }

        fun startActivity(context: Context, sessionId: String, sessionStartDateInMillis: Long,
                          sessionEndDateInMillis: Long, roomId: String?) {
            val intent = Intent(context, SessionDetailActivity::class.java)
            intent.putExtra(PARAM_SESSION_ID, sessionId)
            intent.putExtra(PARAM_SESSION_START_DATE, sessionStartDateInMillis)
            intent.putExtra(PARAM_SESSION_END_DATE, sessionEndDateInMillis)
            intent.putExtra(PARAM_SESSION_ROOM, roomId)
            context.startActivity(intent)
        }
    }

}